webrtc/stun/stun_protocol.nim

```diff
@@ -0,0 +1,155 @@
+# +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+# |                                                               |
+# |                     Transaction ID (96 bits)                  |
+# |                                                               |
+# +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+# Message type:
+#   0x0001: Binding Request
+#   0x0101: Binding Response
+#   0x0111: Binding Error Response
+#   0x0002: Shared Secret Request
+#   0x0102: Shared Secret Response
+#   0x0112: Shared Secret Error Response
+
+  RawStunMessage = object
+    msgType: uint16
+    length* {.bin_value: it.content.len().}: uint16
+    magicCookie: uint32
+    transactionId: array[12, byte] # Down from 16 to 12 bytes in RFC5389
+    content* {.bin_len: it.length.}: seq[byte]
+
+  StunMessage* = object
+    msgType*: uint16
+    transactionId*: array[12, byte]
+    attributes*: seq[RawStunAttribute]
+
+proc getAttribute(attrs: seq[RawStunAttribute], typ: uint16): Option[seq[byte]] =
+  for attr in attrs:
+    if attr.attributeType == typ:
+      return some(attr.value)
+  return none(seq[byte])
+
+proc addLength(msgEncoded: var seq[byte], toAddLength: uint16) =
+  # Add length to an already encoded message. It is necessary because
+  # some attributes (such as Message Integrity or Fingerprint) need
+  # the encoded message to be computed.
+  let
+    currentLength: uint16 = msgEncoded[2].uint16 * 256'u16 + msgEncoded[3].uint16
+    totalLength = currentLength + toAddLength
+  if totalLength < currentLength:
+    raise newException(WebRtcError, "Stun - Try to encode a message larger than uint16 max")
+  msgEncoded[2] = (totalLength div 256'u16).uint8
+  msgEncoded[3] = (totalLength mod 256'u16).uint8
+
+proc decode*(T: typedesc[StunMessage], msg: seq[byte]): StunMessage =
+  let smi = Binary.decode(msg, RawStunMessage)
+  return T(msgType: smi.msgType,
+           transactionId: smi.transactionId,
+           attributes: RawStunAttribute.decode(smi.content))
+
+proc encode*(msg: StunMessage, userOpt: Option[seq[byte]] = none(seq[byte])): seq[byte] =
+  const pad = @[0, 3, 2, 1]
+  var smi = RawStunMessage(msgType: msg.msgType,
+                           magicCookie: StunMagicCookie,
+                           transactionId: msg.transactionId)
+  for attr in msg.attributes:
+    smi.content.add(Binary.encode(attr))
+    smi.content.add(newSeq[byte](pad[smi.content.len() mod 4]))
+
+  result = Binary.encode(smi)
+
+  if userOpt.isSome():
+    let username = string.fromBytes(userOpt.get())
+    let usersplit = username.split(":")
+    if usersplit.len() == 2 and usersplit[0].startsWith("libp2p+webrtc+v1/"):
+      result.addLength(24)
+      result.add(Binary.encode(MessageIntegrity.encode(result, toBytes(usersplit[0]))))
+
+  result.addLength(8)
+  result.add(Binary.encode(Fingerprint.encode(result)))
+
+proc getBindingResponse*(
+    msg: seq[byte],
+    ta: TransportAddress
+  ): Option[seq[byte]] =
```

Comment:
I'd use an `Option` as a return type when it is a normal execution not having anything to return, like trying to get a value from a map, but it doesn't contain it. In this case, is it ever expected to not return a response? It doesn't seem so. Not being able to decode a message is an error. How about changing the `msg` to `StunMessage` instead and raising an exception when something unexpected (error) happens?

Reply:
I reworked this part.
There's now two functions called back to back.
First [checkForError](https://github.com/status-im/nim-webrtc/pull/9/files#diff-5337d70c0d648e8db7070f22c850c47b5012039666d93ced99c62ff59cbc1a54R81) which returns an option with a `StunMessage` if there's an error while in the received message.
Then, if checkForError's result is none, [getBindingResponse](https://github.com/status-im/nim-webrtc/pull/9/files#diff-5337d70c0d648e8db7070f22c850c47b5012039666d93ced99c62ff59cbc1a54R45) which returns the appropriate response (there's, in theory, no error possible at this point).